### ISSUE TITLE
IPv6 Compliance RFC4191: Processing Router Advertisement with Route Information Option (Host Only) [v6LC.2.2.23 Part G]

### ISSUE THREAD

**Issue by LiveFreeAndRoam** [systemd/systemd#31606]
### systemd version the issue has been seen with

255

### Used distribution

Ubuntu 23

### Linux kernel version used

6.5.0-17-generic

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

### Expected behaviour you didn't see

> _**NOTE**: Related is https://github.com/systemd/systemd/issues/28449, which is an issue on 2.2.23 part C. This defect is for 2.2.23 part G._

#### Test v6LC.2.2.23: Stable addresses (Host Only) - Part G: RIO with prefix len of 0 and pref High.
**Purpose**: Verify that the HUT uses a Route Information Options to choose the next-hop.
**Reference**: [RFC 4191, Section 3.1](https://www.ietf.org/rfc/rfc4191.html#section-3.1)

#### Network Configuration
```
       TN2
        | 
    +---+---+
    |       |
   TR1     TR2
    |       |
    +---+---+
        |
       HUT
```

#### Procedure

| Step | Action | Expected Behavior of HUT<sup>(1)</sup> |
|--------|--------|--------|
| 1 | TR1 sends `RA rtr-pref=Medium`. | |
| 2 | TR2 sends RA RIO: len=1 prefixlen=0 prf=High Life=30 prefix=:: |
| 3 | TR1 forwards an Echo Request from TN2 with a source address of 2001:2:0:2000::2 to the HUT. | The HUT should respond to the Echo Request with an Echo Reply with a next-hop of TR2. |

### Unexpected behaviour you saw

In Step 3, the HUT sent an echo reply via TR1 when it should have sent it to TR2, because TR2 announced that it's router preference was High compared to TR1's Medium. Networkd evidently ignores the router preference in TR2's RIO.

### Steps to reproduce the problem

[v6LC_2_2_23g.zip](https://github.com/systemd/systemd/files/14478789/v6LC_2_2_23g.zip)

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
#
# RA from TR1.
#
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 10min
enp0s9: NDISC: Invoking callback for 'router' event.
Setting '/proc/sys/net/ipv6/neigh/enp0s9/base_reachable_time_ms' to '30000'
No change in value '30000', suppressing write
Setting '/proc/sys/net/ipv6/neigh/enp0s9/retrans_time_ms' to '1000'
No change in value '1000', suppressing write
Setting '/proc/sys/net/ipv6/conf/enp0s9/hop_limit' to '64'
No change in value '64', suppressing write
enp0s9: Requesting NDisc route (n/a): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc address (n/a): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: Setting SLAAC addresses and router.
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: Configuring NDisc route (requesting): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Configuring NDisc route (requesting): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Configuring NDisc address (requesting): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: Received new NDisc route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a

#
# RA from TR2. The RIO was ignored.
#
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 10min
enp0s9: NDISC: Invoking callback for 'router' event.
Setting '/proc/sys/net/ipv6/neigh/enp0s9/base_reachable_time_ms' to '30000'
No change in value '30000', suppressing write
Setting '/proc/sys/net/ipv6/neigh/enp0s9/retrans_time_ms' to '1000'
No change in value '1000', suppressing write
Setting '/proc/sys/net/ipv6/conf/enp0s9/hop_limit' to '64'
No change in value '64', suppressing write
enp0s9: Requesting NDisc route (n/a): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Route prefix is ::/0, ignoring
enp0s9: Setting SLAAC addresses and router.
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: Configuring NDisc route (requesting): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: ndisc_check_ready(): SLAAC addresses and routes are not set.
enp0s9: Received new NDisc route (configured): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: ndisc_check_ready(): SLAAC addresses and routes are not set.
enp0s9: Received new NDisc address (configured): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: tentative,manage-temporary-address,no-prefixroute, scope: global
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: ndisc_check_ready(): SLAAC addresses and routes are not set.
enp0s9: Received new NDisc route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Received remembered NDisc route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: ndisc_check_ready(): no SLAAC address is ready.
enp0s9: Received updated NDisc address (configured): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 13s, preferred for 18h 12min 13s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: SLAAC addresses and routes set.
enp0s9: link_check_ready(): IPv4LL:no DHCPv4:no DHCPv6:no DHCP-PD:no NDisc:yes
enp0s9: State changed: configuring -> configured
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=28 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=29 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=30 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
enp0s9: Received new foreign route (configured): dst: 2001:2:0:1000:a00:27ff:fe1a:432/128, src: n/a, gw: n/a, prefsrc: n/a, table: local(255), priority: 0, proto: kernel, scope: global, type: local, flags: n/a

#
# RA from TR1 - Cleanup
#
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 0
enp0s9: NDISC: Invoking callback for 'router' event.
enp0s9: Removing NDisc route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Found an existing route that conflicts with new route based on a received RA, removing.
enp0s9: Removing NDisc route (configured): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc route (configured,removing): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Setting SLAAC addresses and router.
enp0s9: Refuse to configure NDisc route with zero lifetime.
enp0s9: Forgetting removed NDisc route (n/a): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1060: Router=yes, Solicited=no, Override=yes
enp0s9: NDISC: Invoking callback for 'neighbor' event.
enp0s9: Forgetting removed NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a

#
# RA from TR2 - Cleanup
#
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 0
enp0s9: NDISC: Invoking callback for 'router' event.
enp0s9: Removing NDisc route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Route prefix is ::/0, ignoring
enp0s9: Setting SLAAC addresses and router.
enp0s9: State changed: configured -> configuring
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=31 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: Refuse to configure NDisc route with zero lifetime.
enp0s9: Forgetting removed NDisc route (n/a): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1061: Router=yes, Solicited=no, Override=yes
enp0s9: NDISC: Invoking callback for 'neighbor' event.
```

**Comment by LiveFreeAndRoam** [comment]
**NOTE**: This is blocking Test 2.2.23j. Route Information Option with a Prefix Length of 0 and Lifetime set to 0. Once this one is passing, I'll rerun 2.2.23j.

**Comment by LiveFreeAndRoam** [comment]
This stest is now passing.  Interestingly, [2.2.23c](https://github.com/systemd/systemd/issues/28449) is now failing, where previously it was passing. That will require deeper analysis.

NOTE: I'll be away for a week and won't have time to investigate further at the moment. Hopefully, you can determine why 2.2.23c is now failing.

**Comment by LiveFreeAndRoam** [comment]
I reran tests and both 2.2.23c and 2.2.23g are passing.

This case can be closed.

The version tested was:

```
$ git log -n 1
commit 59b72d6f47160c74a1d7f44bc3e2373b4e14dd58 (HEAD -> network-next, origin/network-next)
Author: Yu Watanabe <watanabe.yu+github@gmail.com>
Date:   Sat Mar 16 02:32:50 2024 +0900
```

**Comment by yuwata** [comment]
> I reran tests and both 2.2.23c and 2.2.23g are passing.

Really? I think I have not changed anything related... Strange... I will check.

**Comment by LiveFreeAndRoam** [comment]
The test suite is a little finicky, so I will also revalidate my results when I get back from Easter holidays, next Tue.

**Comment by yuwata** [comment]
Yes, of course. Please enjoy holidays.

**Comment by LiveFreeAndRoam** [comment]
@yuwata, the test still passes.  I successfully reran the test using your fork network-next branch, as well as the upstream main branch:

```
$ git remote -v
origin  https://github.com/yuwata/systemd.git (fetch)
origin  https://github.com/yuwata/systemd.git (push)
upstream        https://github.com/systemd/systemd.git (fetch)
upstream        https://github.com/systemd/systemd.git (push)
$ git log -n 1
commit 59b72d6f47160c74a1d7f44bc3e2373b4e14dd58 (HEAD -> network-next, origin/network-next)
```

```
$ git remote -v
origin  https://github.com/systemd/systemd (fetch)
origin  https://github.com/systemd/systemd (push)
$ git log -n 1
commit 52c15e9c9b97cebd3e2c78393e7c78fd5a213d22 (HEAD -> main, origin/main, origin/HEAD)
```

**Comment by LiveFreeAndRoam** [comment]
I retested 2.2.23c and 2.2.23g, using the same commit IDs I mentioned in the above post, and both are failing again. I also reran using this [commit ID](https://github.com/systemd/systemd/issues/31842#issuecomment-2008310900), where I first observed 2.2.23c passing.  It's failing with that one too.

I'm digging in to see if I can locate the cause of these different results. I have recently rebooted my VM, so that may be related.

For the time-being, we'll have to mark 2.2.23g (this issue) as unresolved and I may need to create a new issue for 2.2.23c.

**Comment by yuwata** [comment]
@LiveFreeAndRoam Thank you!

**Comment by LiveFreeAndRoam** [comment]
@yuwata, no rush, I'm just wanting to understand the process for when things get merged back to systemd/systemd upstream main branch. I periodically test against it, but if you have a plan for merging to the upstream then I'll align with that. If not, then I'll continue with checking the upstream for your merges.

**Comment by yuwata** [comment]
> @yuwata, no rush, I'm just wanting to understand the process for when things get merged back to systemd/systemd upstream main branch. I periodically test against it, but if you have a plan for merging to the upstream then I'll align with that. If not, then I'll continue with checking the upstream for your merges.

@LiveFreeAndRoam Our basic rule about reviewing and merging PR is simple. Each PR needs approval form at least one of the maintainers of systemd.
Unfortunately reviewing process is typically very slow, especially if a PR has too many commits. But opening small PRs may cause conflict and build or test may be failed after merging them. So, I pushed all unreviewed commits into network-next to make each commit not conflict or interfere each other, and pick several commits from the branch and open only a few small PRs to wait for review.
So, when you test my commits, typically it is enough to test network-next branch, and not necessary do that for each small PR.

In short,
- network-next branch: stack of unreviewed commits, for testing and CIs, not ready for review,
- small PRs: ready for review.
